src/librustc_mir_build/hair/pattern/const_to_pat.rs

```diff
@@ -257,7 +238,92 @@ impl<'a, 'tcx> ConstToPat<'a, 'tcx> {
                 slice: None,
                 suffix: Vec::new(),
             },
-            _ => PatKind::Constant { value: cv },
+            // If the type is not structurally comparable, just emit the constant directly, causing
+            // the pattern match code to treat it opaquely.
+            _ if !self.type_marked_structural(cv.ty) => PatKind::Constant { value: cv },
```

Comment:
I did a major change to the previous state of this PR, namely that types not marked as structural are used as constants directly, making them opaque constants (except for various hacks around `&[u8]`, which I hope we can remove after https://github.com/rust-lang/rust/pull/70759

Reply:
This has a problem though: it forces suboptimal type-based errors later, while emitting errors here is better UX IMO.

Reply:
I didn't see anything in the tests, what would you be expecting?

Reply:
If the type isn't structural, and this code doesn't emit errors itself, *the caller* emits the errors later. So instead of specific errors, you just get blanket errors about the whole `const` type.

Reply:
This is preexisting behaviour (the previous code unconditionally returned an opaque constant directly). If we want to add any new diagnostics, we should do so in a separate PR, although I still don't understand what kind of errors the caller would emit.

Reply:
Sorry, I should've just linked the code, this is the fallback: https://github.com/rust-lang/rust/blob/db4de757a7b7d49d05dbd1a58c30ff568c920670/src/librustc_mir_build/hair/pattern/const_to_pat.rs#L99-L101.

> This is preexisting behaviour (the previous code unconditionally returned an opaque constant directly)

I guess, that makes sense. Maybe leave some comments indicating that it's suboptimal?